rust-src/bulletproofs/src/inner_product_proof.rs

```diff
@@ -331,6 +335,118 @@ pub fn verify_inner_product<C: Curve>(
     P_prime.minus_point(&RHS).is_zero_point()
 }
 
+/// This function is an optimized variant of the above.
+/// It is verified whether P'=<a,G>+<b,H'>+<a,b>Q for P' =
+/// multiexp(P_prime_bases, P_prime_exponents) and H'_i =
+/// H_i^H_exponents_i. Arguments:
+/// - transcript - the proof transcript
+/// - gens - generators containing vectors G and H both of length n
+/// - H_exponents - slice of scalars to whose powers the H_i are raised
+/// - P_prime_bases - slice of points for computing curve point P'. It is
+///   assumed that the the first base points are G | H, which are implicit and
+///   omitted from P_prime_bases
+/// - P_prime_exponents - slice of scalars to whose powers the elements
+///   P_prime_bases are raised
+/// - Q - the elliptic curve point Q
+/// - proof - the inner product proof
+/// Precondictions:
+/// G, H, and H_exponents should all be of the same length n, and this
+/// length must a power of 2. Furthermore, the length of P_prime_exponents is
+/// equal to the length of P_prime_bases plus of 2n (since G and H are omitted
+/// from P_prime_bases).
+#[allow(non_snake_case)]
+pub fn verify_inner_product_with_scalars<C: Curve>(
+    transcript: &mut RandomOracle,
+    gens: &Generators<C>,
+    H_exponents: &[C::Scalar],
+    P_prime_bases: &[C],
+    P_prime_exponents: &[C::Scalar],
+    Q: &C,
+    proof: &InnerProductProof<C>,
+) -> bool {
+    let n = gens.G_H.len();
+    let L_R = &proof.lr_vec;
+    let a = proof.a;
+    let b = proof.b;
+    let mut ab = a;
+    ab.mul_assign(&b);
+
+    let verification_scalars = match verify_scalars(transcript, n, proof) {
+        None => return false,
+        Some(scalars) => scalars,
+    };
+    let (u_sq, u_inv_sq, s) = (
+        verification_scalars.u_sq,
+        verification_scalars.u_inv_sq,
+        verification_scalars.s,
+    );
+
+    // nsum = sum^-1 = prod_j L_R[j].0 ^ -u_sq[j] * L_R[j].1 ^ -u_inv_sq[j]
+    let mut nsum_bases = Vec::with_capacity(2 * L_R.len());
+    let mut nsum_exps = Vec::with_capacity(2 * L_R.len());
+    for j in 0..L_R.len() {
+        nsum_bases.push(L_R[j].0);
+        nsum_bases.push(L_R[j].1);
+        let mut musq = u_sq[j];
+        let mut muisq = u_inv_sq[j];
+        musq.negate();
+        muisq.negate();
+        nsum_exps.push(musq);
+        nsum_exps.push(muisq);
+    }
+
+    // RHS = prod_i G_i^(s_i*a) H_i^(H_exponents_i * s_i^-1 * b) * Q^{ab} * nsum
+    let mut s_inv = s.clone();
+    s_inv.reverse();
+    let mut G_exps = s;
+    for ge in &mut G_exps {
+        ge.mul_assign(&a);
+    }
+
+    let mut rhs_bases = Vec::with_capacity(2 * n + 1 + nsum_bases.len() + P_prime_bases.len());
+    // Add G and H from gens to rhs_bases
+    for i in 0..n {
+        rhs_bases.push(gens.G_H[i].0);
+    }
+    for i in 0..n {
+        rhs_bases.push(gens.G_H[i].1);
+    }
+
+    // add further elements to rhs_bases
+    rhs_bases.push(*Q);
+    rhs_bases.append(&mut nsum_bases);
+    rhs_bases.extend(P_prime_bases);
```

Comment:
It seems that Q = g_hat is already in `P_prime_bases` (at least where we are calling `verify_inner_product_with_scalars` from). Is it intentional that we are doing `Q^(ab) Q^(-tx)` and not `Q^(ab-tx)`? It probably does not matter much efficiency wise, just wondering.

Reply:
The only reason is that we may want to call this function also if `P_prime_bases` is different and assuming it contains a particular element in a specific position would make the function less general.

Reply:
I see.

Reply:
Do you need this in the set non-membership proof? Or why is the generality useful @chrmatt ?

Generally one should wait with introducing generality until it is needed.

Reply:
The generality is not needed for the non-membership. But it's not really introducing generality, but rather skipping an optimization: We currently already say that "It is assumed that the first base points are `G | H`, which are implicit and omitted from `P_prime_bases`", which allows a substantial speedup, but does not provide the nicest interface. We would now need to additional assume that `Q` is also implicitly there. We can do that if you think it makes sense, but it seems to overcomplicate the usage of the function.

Reply:
This is (or rather should be) an internal function in this crate.

In that case a complicated interface is fine. I don't particularly mind if it stays like this, but I was replying to the statement you made about "more general". 

But we should make this function `pub(crate)`. I don't think we need it outside the bulletproofs.

Reply:
Made this function `pub(crate)` and added the Q to the implicit bases.